core/src/main/java/org/parchmentmc/feather/mapping/ImmutableMappingDataContainer.java

```diff
@@ -246,18 +249,36 @@ public List<String> getJavadoc() {
             return javadoc;
         }
 
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public ConstantData getConstant() {
+            return constant;
+        }
+
         @Override
         public boolean equals(Object o) {
             if (this == o) return true;
-            if (!(o instanceof FieldData)) return false;
-            FieldData that = (FieldData) o;
-            return getName().equals(that.getName()) && Objects.equals(getDescriptor(), that.getDescriptor())
-                    && getJavadoc().equals(that.getJavadoc());
+            if (!(o instanceof ImmutableFieldData)) return false;
+
+            ImmutableFieldData that = (ImmutableFieldData) o;
+
+            if (getName() != null ? !getName().equals(that.getName()) : that.getName() != null) return false;
+            if (getDescriptor() != null ? !getDescriptor().equals(that.getDescriptor()) : that.getDescriptor() != null)
+                return false;
+            if (getJavadoc() != null ? !getJavadoc().equals(that.getJavadoc()) : that.getJavadoc() != null)
+                return false;
+            return getConstant() != null ? getConstant().equals(that.getConstant()) : that.getConstant() == null;
         }
 
         @Override
         public int hashCode() {
-            return Objects.hash(getName(), getDescriptor(), getJavadoc());
+            int result = getName() != null ? getName().hashCode() : 0;
+            result = 31 * result + (getDescriptor() != null ? getDescriptor().hashCode() : 0);
+            result = 31 * result + (getJavadoc() != null ? getJavadoc().hashCode() : 0);
+            result = 31 * result + (getConstant() != null ? getConstant().hashCode() : 0);
+            return result;
```

Comment:
Why change from `Objects.hash`? Are there any substantial benefits to hardcoding ourselves to this pattern of hashcoding? (I'd also like a source on how this pattern was decided.)

Same comment on all other similar instances.